geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/vlm/RasterSource.scala

```diff
@@ -150,6 +256,131 @@ object RasterSource {
     SpatialTiledRasterLayer(zoom, contextRDD)
   }
 
+  def readOrderedToLayout(
+    sc: SparkContext,
+    layerType: String,
+    paths: java.util.ArrayList[java.util.HashMap[String, String]],
+    layoutType: GPSLayoutType,
+    targetCRS: String,
+    numPartitions: Integer,
+    resampleMethod: ResampleMethod,
+    readMethod: String
+  ): SpatialTiledRasterLayer = {
+    val scalaPaths: Seq[Seq[(String, String)]] = paths.asScala.toSeq.map { _.asScala.toSeq }
+
+    val partitions =
+      numPartitions match {
+        case i: Integer => Some(i.toInt)
+        case null => None
+      }
+
+    readOrderedToLayout(
+      sc,
+      layerType,
+      sc.parallelize(scalaPaths, partitions.getOrElse(sc.defaultParallelism)),
+      layoutType,
+      targetCRS,
+      resampleMethod,
+      readMethod
+    )
+  }
+
+  def readOrderedToLayout(
+    sc: SparkContext,
+    layerType: String,
+    rdd: RDD[Seq[(String, String)]],
+    layoutType: LayoutType,
+    targetCRS: String,
+    resampleMethod: ResampleMethod,
+    readMethod: String
+  ): SpatialTiledRasterLayer = {
+    val rasterSourcesRDD: RDD[Seq[(String, RasterSource)]] =
+      (readMethod match {
+        case GEOTRELLIS =>
+          rdd.mapPartitions { iter =>
+            iter.map { files: Seq[(String, String)] =>
+              files.map { case (k, v) => (k, new GeoTiffRasterSource(v): RasterSource)
+              }
+            }
+          }
+        case GDAL =>
+          rdd.mapPartitions { iter =>
+            iter.map { files: Seq[(String, String)] =>
+              files.map { case (k, v) => (k, GDALRasterSource(v): RasterSource)
+              }
+            }
+          }
+      }).cache()
+
+    val orderedRasterSourcesRDD: RDD[Seq[RasterSource]] =
+      rasterSourcesRDD.map { keyedSource: Seq[(String, RasterSource)] =>
+        keyedSource.sortBy { _._1 }.map { case (_, source) => source }
```

Comment:
I assume by the test that the input is `[(bandOrder, source), ..]`.

For the base use case where you need to ingest 1000 (lets say) Landsat senes you're going to have 1000 sense x 3 bands = 3000 entries in essentially random order. When `sc.parallelize` happens you have no way to know how they were grouped which means any sorting per partition is now partial. You may have gotten entries where all of the bands are "1" or "3" or maybe bands "1", "2", "3" but for scenes that do not spatially overlap.

Reply:
Let's say there are two Landsat scenes, three bands each `(a, b, c)` where all bands for scene 1 and scene 2 are on different partitions. However they both overlap same target tile `SpatialKey(x, y)` in the destination layout. Without a shuffle there are going to be two resulting tiles for `SpatialKey(x, y)`. For things to be in expected format of `TiledRasterSource` they need to be merged I believe.

![img_2666](https://user-images.githubusercontent.com/1158084/50015005-7ac7f400-ff93-11e8-9408-6dfec1a6fbae.jpg)

An additional problem is that even restriction each partition to single scene it's easy to read data sizes that are much larger than optimal. All the bands of LC8 are going to be well over a gigabyte in memory when optimal partition sizes appears to be one around 200MB.